pkg/ingestor/parser/vuln/vuln.go

```diff
@@ -0,0 +1,167 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The Vulnerability attestation parser parses the attestation defined by by
+// the certifier using the predicate type
+// "https://in-toto.io/attestation/vuln/v0.1" Based on the information
+// contained, a package node is generated with just purl and digest (if
+// found). This package node is merged into the graph database as it already
+// exists.
+//
+// An attestation node is generated that contains the information stored in the
+// attestation document.  The attestation node is linked to the package node
+// via a "attestation" edge type.
+//
+// Depending on the number of vulnerabilities found, vulnerability nodes are
+// generated that contain just the vulnerability ID that can be used to query
+// for more information as needed.  The vulnerability node is linked to the
+// attestation node via a "vulnerable" edge type.
+package vuln
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/guacsec/guac/pkg/assembler"
+	"github.com/guacsec/guac/pkg/assembler/clients/generated"
+	"github.com/guacsec/guac/pkg/assembler/helpers"
+	attestation_vuln "github.com/guacsec/guac/pkg/certifier/attestation"
+	"github.com/guacsec/guac/pkg/handler/processor"
+	"github.com/guacsec/guac/pkg/ingestor/parser/common"
+)
+
+type parser struct {
+	packages []*generated.PkgInputSpec
+	vulnData *generated.VulnerabilityMetaDataInput
+	vulns    []*generated.OSVInputSpec
+	isVulns  []assembler.IsVulnIngest
+}
+
+// NewVulnCertificationParser initializes the parser
+func NewVulnCertificationParser() common.DocumentParser {
+	return &parser{}
+}
+
+// Parse breaks out the document into the graph components
+func (c *parser) Parse(ctx context.Context, doc *processor.Document) error {
+	statement, err := parseVulnCertifyPredicate(doc.Blob)
+	if err != nil {
+		return fmt.Errorf("failed to parse slsa predicate: %w", err)
+	}
+	ps, err := parseSubject(statement)
+	if err != nil {
+		return fmt.Errorf("unable to parse subject of statement: %w", err)
+	}
+	c.packages = ps
+	c.vulnData = parseMetadata(statement, doc)
+	vs, ivs, err := parseVulns(statement, doc)
+	if err != nil {
+		return fmt.Errorf("unable to parse vulns of statement: %w", err)
+	}
+	c.vulns = vs
+	c.isVulns = ivs
+	return nil
+}
+
+func parseVulnCertifyPredicate(p []byte) (*attestation_vuln.VulnerabilityStatement,
+	error) {
+	predicate := attestation_vuln.VulnerabilityStatement{}
+	if err := json.Unmarshal(p, &predicate); err != nil {
+		return nil, err
+	}
+	return &predicate, nil
+}
+
+func parseSubject(s *attestation_vuln.VulnerabilityStatement) ([]*generated.PkgInputSpec,
+	error) {
+	var ps []*generated.PkgInputSpec
+	for _, sub := range s.StatementHeader.Subject {
+		p, err := helpers.PurlToPkg(sub.Name)
+		if err != nil {
+			return nil, fmt.Errorf("bad purl in statement header: %w", err)
+		}
+		ps = append(ps, p)
+		// FIXME Do we do anything with Digest here?
```

Comment:
if there is a digest, you can create an artifact that you can link to the created package above (via the purl) using `IsOccurrence`

Reply:
SG